### PR TITLE
fix: converting between local and remote [INS-3299]

### PR DESCRIPTION
**Pull request by marckong**
### Description
In this PR, converting between the local and cloud projects, we are duplicating the project and its files with new unique identifiers and delete the original ones. When selecting the cloud project, we are also automatically syncing to the cloud, which is the expected behavior in the product requirement.

### Steps to test
1. sign into the app on staging
2. create local and cloud projects
3. invite a member
4. convert them in between local and cloud
5. confirm the cloud projects (converted between) are correctly loading all the files

### CODE REVIEW

**Comment by jackkav** [comment]
This creates a new problem where workspaces can potentially be orphaned but we only have a recovery path for projects.

**Comment by marckong** [comment]
@jackkav can you please elaborate more and which path are we orphaning the workspace (I suppose you meant as file)? database.removeWhere actually deletes all its descendants. If it is not deleted, obviously this conversion failed, but the app still have duplicated the records, so users haven't lost any data.

**Comment by marckong** [comment]
Converting between local and remote requires root_document_id to be refreshed. NeDB does not allow _id update, which is expected. Dupilcating the record will inherit all the descendants with new unique identifiers.

**Review by ihexxa — COMMENTED** [review]
Tested together with @marckong and this patch overall works well, is it possible perform a test to resolve jack's concern?

**Review by filfreire — APPROVED** [review]
LGTM, tested this alongside @marckong 

Seemed to work ok locally.

We found 2 minor edge-cases, unrelated to this PR - INS-3307 and INS-3306

**Review by ihexxa — APPROVED** [review]
lgtm+1

File: packages/insomnia/src/ui/routes/actions.tsx [hunk 1]
```diff
@@ -182,8 +186,25 @@ export const updateProjectAction: ActionFunction = async ({
...
+      const newRemoteProject = await database.duplicate(project, { remoteId: newCloudProject.id });
+      const vcs = VCSInstance();
+      if (vcs) {
+        const filesForSync = await database.find<Workspace>(models.workspace.type, { parentId: newRemoteProject._id });
+
+        for (const fileForSync of filesForSync) {
+          const workspaceMeta = await models.workspaceMeta.getOrCreateByParentId(fileForSync._id);
+          if (!workspaceMeta.gitRepositoryId) {
+            await initializeLocalBackendProjectAndMarkForSync({ vcs, workspace: fileForSync });
+            await pushSnapshotOnInitialize({ vcs, workspace: fileForSync, project: newRemoteProject });
```
Comment (gatzjames) [inline packages/insomnia/src/ui/routes/actions.tsx:200]: This goes over the wire so it is possible to fail.
We need to figure out what should happen in the error case before merging this

File: packages/insomnia/src/ui/routes/actions.tsx [hunk 2]
```diff
@@ -182,8 +186,25 @@ export const updateProjectAction: ActionFunction = async ({
...
+          }
+        }
+      }
+
+      await database.removeWhere(project.type, { _id: project._id });
+      await database.flushChanges(flushId);
+      return redirect(`/organization/${organizationId}/project/${newRemoteProject._id}`);
```
Comment (jackkav) [inline packages/insomnia/src/ui/routes/actions.tsx:207]: I see no reason to add buffer and flush changes to either of these actions.

File: packages/insomnia/src/ui/routes/actions.tsx [hunk 3]
```diff
@@ -182,8 +186,25 @@ export const updateProjectAction: ActionFunction = async ({
         };
       }
 
-      await models.project.update(project, { name, remoteId: newCloudProject.id });
-      return null;
+      const flushId = await database.bufferChanges();
+
+      const newRemoteProject = await database.duplicate(project, { remoteId: newCloudProject.id });
```
Comment (jackkav) [inline packages/insomnia/src/ui/routes/actions.tsx:191]: If a user wants to convert from local to cloud and this breaks request chains, I don't think product would find this acceptable.

Comment (marckong) [reply packages/insomnia/src/ui/routes/actions.tsx:191]: This happens when the user clicks Duplicate button too. We should fix it in a different PR and out of this PR's scope. The more important part is to fix cloud <-> local <-> cloud conversion that breaks the app as the files are not in the correct team project. Root document id needs to change. The product confirmed that we won't change this conversion feature for 8.4. What do you suggest?
